secure-device-storage/sds-android/securedevicestorage/src/main/java/de/adorsys/android/securedevicestorage/KeystoreTool.java

```diff
@@ -0,0 +1,250 @@
+package de.adorsys.android.securedevicestorage;
+
+import android.content.Context;
+import android.os.Build;
+import android.security.KeyPairGeneratorSpec;
+import android.security.keystore.KeyGenParameterSpec;
+import android.security.keystore.KeyProperties;
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.support.annotation.RequiresApi;
+import android.util.Base64;
+import android.util.Log;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.math.BigInteger;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.KeyPairGenerator;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.UnrecoverableEntryException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+import java.util.ArrayList;
+import java.util.Calendar;
+
+import javax.crypto.Cipher;
+import javax.crypto.CipherInputStream;
+import javax.crypto.CipherOutputStream;
+import javax.crypto.NoSuchPaddingException;
+import javax.security.auth.x500.X500Principal;
+
+import static android.os.Build.VERSION_CODES.M;
+
+public class KeystoreTool {
+    private static final String KEY_ALIAS = "adorsysKeyPair";
+    private static final String KEY_ENCRYPTION_ALGORITHM = "RSA";
+    private static final String KEY_KEYSTORE_NAME = "AndroidKeyStore";
+    private static final String KEY_CIPHER_JELLYBEAN_PROVIDER = "AndroidOpenSSL";
+    private static final String KEY_CIPHER_MARSHMALLOW_PROVIDER = "AndroidKeyStoreBCWorkaround";
+    private static final String KEY_TRANSFORMATION_ALGORITHM = "RSA/ECB/PKCS1Padding";
+    private static final String KEY_X500PRINCIPAL = "CN=SecureDeviceStorage, O=Adorsys, C=Germany";
+
+    public static boolean keyPairExists() {
+        try {
+            return getKeyStoreInstance().getKey(KEY_ALIAS, null) != null;
+        } catch (KeyStoreException | NoSuchAlgorithmException | CertificateException
+                | UnrecoverableKeyException | IOException e) {
+
+            if (BuildConfig.DEBUG) {
+                Log.e(KeystoreTool.class.getName(), e.getMessage(), e);
+            }
+            return false;
+        }
+    }
+
+    static void generateKeyPair(@NonNull Context context)
+            throws InvalidAlgorithmParameterException,
+            NoSuchProviderException, NoSuchAlgorithmException, UnrecoverableKeyException,
+            CertificateException, KeyStoreException, IOException {
+
+        // Create new key if needed
+        if (!keyPairExists()) {
+            if (Build.VERSION.SDK_INT >= M) {
+                generateJellyBeanKeyPair(context);
```

Comment:
Update if-else clause. For now we have to use the JellyBean implementation on all API until we implement the Marshmallow logic.